File: backend/app/routers/topology.py
"""Topology graph — /api/topology."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from pydantic import TypeAdapter
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.db import get_session as get_db
from app.schemas.topology import TopologyResponse
from app.services import topology as service
from app.services.read_cache import cached_read

router = APIRouter(prefix="/topology", tags=["topology"])

# Module level: building a TypeAdapter walks the whole model, which is not
# something to redo on every request.
_RESPONSE_ADAPTER = TypeAdapter(TopologyResponse)


@router.get(
    "", response_model=TopologyResponse, dependencies=[Depends(get_current_user)]
)
async def get_topology(
    site_id: int | None = Query(
        default=None, gt=0, description="Restrict to switches and devices in this site."
    ),
    room_id: int | None = Query(
        default=None, gt=0, description="Restrict to a single room."
    ),
    vlan_id: int | None = Query(
        default=None,
        gt=0,
        description=(
            "Keep only switches carrying this VLAN (native or tagged) on at "
            "least one port. Links between those switches are all returned — "
            "filtering edges too would hide the cable carrying the VLAN."
        ),
    ),
    include_devices: bool = Query(
        default=True,
        description=(
            "Include non-switch devices as leaf nodes, with an `attachment` "
            "edge per port they are plugged into. Turn off for a "
            "switch-only backbone view."
        ),
    ),
    db: AsyncSession = Depends(get_db),
) -> TopologyResponse:
    # Cached when REDIS_URL is set — the SPA refetches the whole graph on every
    # filter toggle. `cached_read` keys on a fingerprint of the inventory, so a
    # write is never invisible here; without Redis this is a direct call.
    return await cached_read(
        db,
        name="topology",
        params={
            "site_id": site_id,
            "room_id": room_id,
            "vlan_id": vlan_id,
            "include_devices": include_devices,
        },
        adapter=_RESPONSE_ADAPTER,
        builder=lambda: service.build_topology(
            db,
            site_id=site_id,
            room_id=room_id,
            vlan_id=vlan_id,
            include_devices=include_devices,
        ),
    )

File: frontend/src/api/endpoints/topology.ts
import { request } from '@/api/client'
import type { TopologyResponse } from '@/api/types'

export interface TopologyQuery {
  /** Restrict to switches and devices in this site. */
  siteId?: number | null
  /** Restrict to a single room. */
  roomId?: number | null
  /**
   * Keep only switches carrying this VLAN (native or tagged) on at least one
   * port. Every link between those switches is still returned.
   */
  vlanId?: number | null
  /**
   * Include non-switch devices as leaf nodes, with an `attachment` edge per
   * port they're plugged into. Off gives a switch-only backbone view.
   */
  includeDevices?: boolean
}

export const topologyApi = {
  /**
   * Returns the graph in Cytoscape element shape: `{ nodes, edges, stats }`.
   * Nodes carry a `parent` id so sites and rooms render as compound group
   * boxes without the client reconstructing the hierarchy.
   */
  get(query: TopologyQuery = {}): Promise<TopologyResponse> {
    const params = new URLSearchParams()
    if (query.siteId) params.set('site_id', String(query.siteId))
    if (query.roomId) params.set('room_id', String(query.roomId))
    if (query.vlanId) params.set('vlan_id', String(query.vlanId))
    // Only sent when explicitly off — the backend default is on, so omitting
    // it keeps the query string short for the common case.
    if (query.includeDevices === false) params.set('include_devices', 'false')
    const qs = params.toString()
    return request<TopologyResponse>({
      method: 'GET',
      url: qs ? `/topology?${qs}` : '/topology',
    })
  },
}
